unity/unity_utils.c

```diff
@@ -198,3 +200,216 @@ unity_mono_method_is_generic (MonoMethod* method)
+			src += length;
+			*result = mono_reflection_type_from_name (utf8Str, image);
+			if (!*result)
+				g_warning ("Cannot load type '%s'", utf8Str);
+			g_free (utf8Str);
+			return src - oldSrc;
+		}
+		default:
+		{
+			g_error ("Type 0x%02x not handled in custom attr value decoding", type);
+			return 0;
+		}
+	}
+}
+
+void unity_mono_unpack_custom_attr_params(MonoImage* image, MonoCustomAttrEntry* entry, void* output, UnityMonoCustomAttrNamedParameterFunc namedParameterHandler)
+{
+	// Mimicing create_custom_attr in mono/metadata/reflection.c for unpacking the params
+	// Also helpful: http://geekswithblogs.net/simonc/archive/2011/06/03/anatomy-of-a-.net-assembly---custom-attribute-encoding.aspx
+	const char *src = (const char*)entry->data;
+	char *dest = (char*)output;
+	int i;
+	guint32 num_named;
+	
+	MonoMethodSignature *sig = mono_method_signature(entry->ctor);
+	
+	// Skip prolog
+	src += 2;
+	for(i = 0; i < sig->param_count; ++i)
+	{
+		MonoType* paramType = sig->params[i];
+		int type = paramType->type;
+		
+		if (type == MONO_TYPE_VALUETYPE && paramType->data.klass->enumtype)
+			type = mono_class_enum_basetype(paramType->data.klass)->type;
+		
+		src += unpack_custom_attr_value(src, type, dest, image);
+		
+		// Advance dest as well
+		switch(type)
+		{
+			case MONO_TYPE_U1:
+			case MONO_TYPE_I1:
+			case MONO_TYPE_BOOLEAN:
+				dest += 1;
+				break;
+			case MONO_TYPE_CHAR:
+			case MONO_TYPE_U2:
+			case MONO_TYPE_I2:
+				dest += 2;
+				break;
+#if SIZEOF_VOID_P == 4
+			case MONO_TYPE_U:
+			case MONO_TYPE_I:
+#endif
+			case MONO_TYPE_R4:
+			case MONO_TYPE_U4:
+			case MONO_TYPE_I4:
+				dest += 4;
+				break;
+#if SIZEOF_VOID_P == 8
+			case MONO_TYPE_U:
+			case MONO_TYPE_I:
+#endif
+			case MONO_TYPE_U8:
+			case MONO_TYPE_I8:
+			case MONO_TYPE_R8:
+				dest += 8;
+				break;
+			case MONO_TYPE_STRING:
+				dest += sizeof(UnityMonoMetadataString*);
+				break;
+			case MONO_TYPE_CLASS:
+				dest += sizeof(MonoType*);
+				break;
+			default:
+				g_error("Unknown/unsupported parameter type");
+				break;
+		}
+	}
+	
+	// Now, named parameters
+	if(namedParameterHandler == NULL) return;
+	num_named = read16(src);
+	src += 2;
+	for (i = 0; i < num_named; ++i)
+	{
+		UnityMonoMetadataString* paramName;
+		int type;
+		// The value will be unpacked into this buffer. The actual value size depends on the type,
+		// but the largest thing we can unpack right now is an int64, so an 8 byte buffer should be sufficient.
+		char buffer[8];
+		
+		src++; // skip the field/property specifier
+		type = *(src++);
+		
+		if (type == MONO_TYPE_SZARRAY)
+		{
+			g_error("Array types are not supported for named parameter unpacking");
+			type = *(src++);
+		}
+		
+		if (type == MONO_TYPE_ENUM)
+		{
+			// Enum typed parameters pack the actual name of the enum type
+			// We don't really care about it for our purposes, just skip past it
+			gint type_len = mono_metadata_decode_blob_size (src, &src);
+			src += type_len;
+		}
+		
```

Comment:
I see. I didn't quite understand that these two cases were the only ones that required special handling. You could perhaps add this as a comment.